### PR TITLE
Refactor and simplify generic printer for unknown objects

### PR DESCRIPTION
**Pull request by smarterclayton**
The first two commits are part of other PRs

@kubernetes/sig-cli-pr-reviews part of the general refactoring for server side print

### CODE REVIEW

**Comment by smarterclayton** [comment]
This simplifies what we added in #44222 - 1.8 will allow the client to request generic printing from the server which reduces the need for this behavior.  This also unifies our unknown printing into a single path.

**Comment by smarterclayton** [comment]
@fabianofranz you're probably the best to review this

**Comment by fabianofranz** [comment]
On it.

File: pkg/printers/internalversion/printers.go [hunk 1]
```diff
@@ -124,20 +112,93 @@ func AddHandlers(h printers.PrintHandler) {
...
+		{Name: "Ready", Type: "integer", Description: extensionsv1beta1.ReplicaSetStatus{}.SwaggerDoc()["readyReplicas"]},
+		{Name: "Age", Type: "string", Description: metav1.ObjectMeta{}.SwaggerDoc()["creationTimestamp"]},
+		{Name: "Containers", Type: "string", Priority: 1, Description: "Names of each container in the template."},
+		{Name: "Images", Type: "string", Priority: 1, Description: "Images referenced by each container in the template."},
+		{Name: "Selector", Type: "string", Priority: 1, Description: extensionsv1beta1.ReplicaSetSpec{}.SwaggerDoc()["selector"]},
+	}
+	h.TableHandler(replicaSetColumnDefinitions, printReplicaSet)
+	h.TableHandler(replicaSetColumnDefinitions, printReplicaSetList)
+
+	daemenSetColumnDefinitions := []metav1alpha1.TableColumnDefinition{
```
Comment (shiywang) [inline pkg/printers/internalversion/printers.go]: daemon

File: pkg/printers/humanreadable.go [hunk 2]
```diff
@@ -44,6 +40,7 @@ type TablePrinter interface {
 type PrintHandler interface {
 	Handler(columns, columnsWithWide []string, printFunc interface{}) error
```
Comment (shiywang) [inline pkg/printers/humanreadable.go]: we already switch all `Handler` to `TableHandler` or `DefaultTableHandler` so this function should be also deprecated right ?

**Comment by fabianofranz** [comment]
@smarterclayton please rebase.

File: pkg/printers/humanreadable.go [hunk 3]
```diff
@@ -299,141 +317,30 @@ func (h *HumanReadablePrinter) PrintObj(obj runtime.Object, output io.Writer) er
...
+			continue
 		}
-
-		// if the error isn't nil, report the "I don't recognize this" error
-		if err := printUnstructured(unstructured, output, discoveredFieldNames, h.options); err != nil {
-			return err
+		lastType, err := printRowsForHandlerEntry(output, handler, obj, h.options, h.lastType)
+		if err != nil {
+			// we try all handlers
+			fmt.Printf("err: %v\n", err)
```
Comment (smarterclayton) [inline pkg/printers/humanreadable.go]: Need to remove this line

File: pkg/printers/internalversion/printers.go [hunk 4]
```diff
@@ -249,23 +241,60 @@ func AddHandlers(h printers.PrintHandler) {
...
+	objectMetaColumnDefinitions := []metav1alpha1.TableColumnDefinition{
+		{Name: "Name", Type: "string", Format: "name", Description: metav1.ObjectMeta{}.SwaggerDoc()["name"]},
+		{Name: "Age", Type: "string", Description: metav1.ObjectMeta{}.SwaggerDoc()["creationTimestamp"]},
+	}
+	h.DefaultTableHandler(objectMetaColumnDefinitions, printObjectMeta)
+}
+
+func printObjectMeta(obj runtime.Object, options printers.PrintOptions) ([]metav1alpha1.TableRow, error) {
+	if meta.IsListType(obj) {
+		fmt.Printf("DEBUG: list type: %#v\n", obj)
```
Comment (smarterclayton) [inline pkg/printers/internalversion/printers.go]: Needs to be removed

**Comment by fabianofranz** [comment]
@smarterclayton This makes generic printing only [NAMESPACE] + NAME + AGE, correct? I'm a little concerned about it for the service catalog since they're using https://github.com/kubernetes/kubernetes/pull/44222 for get and describe, and this will be kind of a regression in functionality until they have their own server-side table printers.

**Comment by smarterclayton** [comment]
Maybe we should work on getting them server side printers then.  My problem
with this code is that's it's generic but effectively random - picking
fields into the columns we have available alphabetically doesn't scale.
It's also not stable - anyone with a field that can be null is going to get
a varying set of columns.  Why shouldn't the user be encouraged to use
custom columns or yaml?   Why guess?



On Jun 29, 2017, at 4:17 PM, Fabiano Franz <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> This makes generic
printing only [NAMESPACE] + NAME + AGE, correct? I'm a little concerned
about it for the service catalog since they're using #44222
<https://github.com/kubernetes/kubernetes/pull/44222> for get and describe,
and this will be kind of a regression in functionality until they have
their own server-side table printers.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/48033#issuecomment-312091739>,
or mute the thread

File: pkg/printers/humanreadable.go (hunk 3 above)
Comment (fabianofranz) [inline pkg/printers/humanreadable.go]: Maybe make it a `glog.V(n)...`

File: pkg/printers/humanreadable.go [hunk 5]
```diff
@@ -631,6 +538,88 @@ func (h *HumanReadablePrinter) PrintTable(obj runtime.Object, options PrintOptio
 	return table, nil
 }
 
+// printRowsForHandlerEntry prints the incremental table output (headers if the current type is
+// different from lastType) including all the rows in the object. It returns the current type
+// or an error, if any.
+func printRowsForHandlerEntry(output io.Writer, handler *handlerEntry, obj runtime.Object, options PrintOptions, lastType reflect.Type) (reflect.Type, error) {
+	t := reflect.TypeOf(obj)
+	if !options.NoHeaders && t != lastType {
```
Comment (fabianofranz) [inline pkg/printers/humanreadable.go]: Does `t != lastType` handle correctly `UnstructuredObject` (with different actual types)?

File: pkg/printers/humanreadable.go [hunk 6]
```diff
@@ -631,6 +538,88 @@ func (h *HumanReadablePrinter) PrintTable(obj runtime.Object, options PrintOptio
...
+		var headers []string
+		for _, column := range handler.columnDefinitions {
+			if column.Priority != 0 && !options.Wide {
```
Comment (fabianofranz) [inline pkg/printers/humanreadable.go]: `--wide` adds all columns?

Comment (smarterclayton) [inline pkg/printers/humanreadable.go]: Yes (this was just code movement, not changes).

File: pkg/printers/humanreadable.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/printers/humanreadable.go]: So I think the right solution here is not to pass individual unstructured down into print obj, but to pass tables down.  And Get should be trying to convert responses from the server into tabular output.  We don't want two different unstructured from a single list to have different columns, and PrintObj can now accept tables.

So as the replacement for the previous dynamic logic, in 1.8 Get will take on the responsibility for turning server lists into Tables and passing them to PrintObj.  That means that the default handler for unstructured won't be used anymore.

**Comment by smarterclayton** [comment]
@fabianofranz if we do the work to change get.go in 1.8 to use Table, then I'll add an adapter for unstructured lists that is generic in this fashion.  I'd prefer to make generic printing of unrecognized resources a higher level concern than this printer, because it will have more context to make the right choice in output.

Any other issues with this PR besides that?

**Comment by fabianofranz** [comment]
Ok, works for me. Just remove the debug prints then this will be good.

**Pull request merged**